### ISSUE TITLE
Name clash for $outer pointers of two different nesting levels

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#2382]
As minimized by @nicolasstucki:

``` scala
object HelloWorld extends js.JSApp {
  def assertEquals(expect: Int, actual: Int, field: String): Unit =
    assert(expect == actual, s"Expected $field to be <$expect> but was <$actual>")

  def main() {
    val b = new B { }
    val y = new b.Y
    val z = new y.Z

    assertEquals(1, b.a, "b.a")
    assertEquals(1, z.z1, "z.z1")
    assertEquals(2, y.y, "y.y")

    assertEquals(2, z.z2, "z.z2")
    //  assertEquals(2, z.z3, "z.z3") // for variant

  }
}

trait A {
  val a: Int = 1
  class X {
    def x: Int = a
  }
}

trait B extends A {
  class Y {
    def y: Int = 2
    class Z extends X {
      def z1: Int = a
      def z2: Int = x
    }
  }
}

//// Variant that also fails
//trait B extends A {
//  class Y {
//    def y: Int = 2
//    class Z extends X {
//      def z1: Int = a
//      def z3: Int = y
//    }
//  }
//}
```

**Comment by nicolasstucki** [comment]
Found out that the second variant can be minimized further to:

```scala
trait A {
  val a: Int = 1
  class X
}

trait B extends A {
  class Y {
    def y: Int = 2
    class Z extends X {
      def z1: Int = a
      def z3: Int = y
    }
  }
}
```

**Comment by nicolasstucki** [comment]
Further minimization to:
```scala
trait A {
  val a: Int = 1
  class X
  class Y {
    class Z extends X {
      def b: Int = a
    }
  }
}
```

**Comment by nicolasstucki** [comment]
With the optimizer disabled the failure shows itself on a missing `$outer` method

```
org.mozilla.javascript.EcmaError: TypeError: Cannot find function helloworld$A$Y$$$outer__Lhelloworld_A in object helloworld.A@1. (Lhelloworld_A$Y$Z.sjsir#14)
```

**Comment by sjrd** [comment]
So the fix in PR #2384 proved to be binary incompatible, so that doesn't work. Consequently, I imagined a link-time hack to disambiguate `$outer` field at deserialization time based on their type:
https://github.com/scala-js/scala-js/compare/master...sjrd:fix-outer-at-link-time
That works well for Scala classes, but it does not work for Scala.js-defined JS classes, because the type of their instances is `any`. So if the *outer* class of a class is a Scala.js-defined JS class, it would still expose the bug. Even though that fixes the scala-compiler use case, it's both too hacky and not good enough to justify its inclusion.

So basically, we're stuck. This fix will have to wait for a binary breaking release, i.e., 1.0.0.

**Comment by sjrd** [comment]
This issue closed automatically, but in fact it's still not fixed for Scala 2.10.x and 2.11.x, so reopening.
